### PR TITLE
Update domain object and repository to reflect how to use SDN for version 4.2

### LINKED ISSUE #235: Update Spring Data Neo4j Example
Update to show use of `Neo4jRepository`.

### PR DESCRIPTION
**Pull request by mangrish**
Closes #235.

### CODE REVIEW

**Comment by mangrish** [comment]
@olivergierke: I haven't updated the VersionEye issues because it's probably not relevant for this change. Let me know if I need to update anything to complete this merge.

File: neo4j/example/src/main/java/example/springdata/neo4j/ActorRepository.java [hunk 1]
```diff
@@ -15,11 +15,15 @@
...
 
 /**
  * {@link GraphRepository} for {@link Actor}s.
- * 
+ *
  * @author Luanne Misquitta
+ * @author Mark Angrish
  */
-public interface ActorRepository extends GraphRepository<Actor> {}
+public interface ActorRepository extends Neo4jRepository<Actor, UUID> {}
```
Comment (olivergierke) [inline neo4j/example/src/main/java/example/springdata/neo4j/ActorRepository.java]: Can you clarify on a couple of things here?

- Does the repository actually need to extend a store-specific interface? We usually recommend not to do so as is exposes store details to its clients.
- You change the ID type to `UUID` but the domain class still contains a `Long` identifier? That is at least confusing but actually even looks broken.

Comment (mangrish) [inline neo4j/example/src/main/java/example/springdata/neo4j/ActorRepository.java]: 1. No it doesn't. In fact until you mentioned it I actually thought it was normal to extend the store specific interface. I'll change this.
1. Argh.. This has been a problem for a while now. The OGM and SDN 3.x both expose a neo4j database implementation called `ID` which currently needs to be annotated with `@GraphId`. Unlike other DB's it's *NOT* guaranteed to be unique across database restarts ([read more here](http://blog.armbruster-it.de/2013/08/assigning-uuids-to-neo4j-nodes-and-relationships/) - it's still relevant). On top of this developers have asked to have a way of uniquely identifying their domain objects, either naturally or with the Neo4j recommended approach of using `UUID`s.  Furthermore the implications of this result in things like: https://jira.spring.io/browse/DATAGRAPH-918 and https://github.com/spring-projects/spring-data-neo4j/issues/366. Ideally this implementation detail will be hidden in future. Personally I'm hoping when we can remove default constructors we can also remove the need for exposing the internal ID. What I will do for now is add some javadoc here to make this clear for those who haven't used Neo4j before.

**Comment by mangrish** [comment]
@olivergierke To make this easier to understand (without the use of the converter) I'm using a string instead.

Commit message wise I was thinking: *"Update domain object and repository to reflect how to use SDN for version 4.2"*. I'm looking [here](https://github.com/spring-projects/spring-data-examples/commits/master) for inspiration here but can't see what else I can add. Any tips?

**Comment by mangrish** [comment]
@olivergierke Updated. I've tried to make the example more clear with the intent and documentation.

**Comment by olivergierke** [comment]
I am still not sure I can follow: we still have a `Long` identifier but the repository exposing a `String` id. That's highly unusual to the average Spring Data developer. `findOne(…)` is supposed to lookup the entity by they identifier. It looks like that's not the case here. Can you elaborate? This seems to be a very fundamental change to what developers expect in a Spring Data setup.

Regarding the commit message: if I read that and look at the change I'd get the impression that I now *need* add that indexed property to all of my domain types to make them work with SDN 4.2. Is that really the case?

**Comment by mangrish** [comment]
@olivergierke I wish we could get rid of the `Long` id but we can't at the moment (for the same reason we can't remove the default constructor). The whole `Long`/`GraphId` is a misnomer; It represents an object in a neo4j database _but_ that id may point to a completely different object or to no object at all on a restart of the database. Consequently it's an encapsulation leak from the database into the OGM and also into SDN.

So we now have the ability to work with people who don't care about database restarts (like those who use the embedded database once off) whereby they can use a repository of ID `Long` (See `GraphRepository`) and those that want to use a natural or synthetic ID which will work across database restarts that want to use a repository of type `Whatever implements Serializable`.

**Comment by mangrish** [comment]
What if we just have two examples: one with what was there previously and one with what we recommend people to use (the one I'm adding now)?

Besides that, I'm a bit confused about what else needs to change exactly for the example. SD/SDN can't abstract the user completely from understanding fundamental Neo4j concepts no?

**Comment by olivergierke** [comment]
I've never been a great fan of all these store details leaking into the domain types. Is there a chance that developers could just use Spring Data's `@Id` and basically get that `@Index(unique = true, primary = true)` stuff automatically?

**Comment by olivergierke** [comment]
What I am wondering, too: if your newly added property *is* the ID, how is the Spring Data `MappingContext` implementation actually finding out about this? There doesn't seem to be anything unique about the property as the `@Index` annotation could be used on other properties, too, right? I am wondering as e.g. Spring Data REST relies on that information to correctly map incoming requests to the actual identifier values.

**Comment by mangrish** [comment]
@olivergierke RE: The idea about `@Id`. Let me run it up the flagpole and see what comes back. I've always liked the idea of `@Id` and I had that concept in the OGM I wrote.

RE: `MappingContext` it's delegated to the OGM via the `Neo4jPersistentProperty`. I have not yet refactored this part of the code to integrate more tightly with Spring Data yet through. The links I sent you above refer to Spring REST's problems with the `id` field.

I'm spread pretty thin at the moment so I'll get to updating it when I can :\

RE: This example. While we have taken some big steps to get closer to Spring Data there are still more to go. At the moment I want to encourage users to use SDN in a way that will allow us to eventually remove the `Long` (or even define it themselves if they run in a one time embedded environment). In prep for 4.2.0 what do you think the example should show?

**Comment by mp911de** [comment]
Since we intend to revamp the Neo4j module entirely, it does not make sense to keep this pull request around.

**Pull request closed without merging**